File: Painting a Grid With Three Different Colors/Painting a Grid With Three Different Colors.py
# Solution for Painting a Grid With Three Different Colors in PY

class Solution:

    def __init__(self):
        self.columnStates = []
        self.MOD = 10**9 + 7
        self.dp = []

    def generateColumnStates(self, curr: str, prevChar: str, l: int, m: int):
        if l == m:
            self.columnStates.append(curr)
            return
        for currColor in {'R', 'G', 'B'}:
            if currColor == prevChar:
                continue
            self.generateColumnStates(curr + currColor, currColor, l + 1, m)

    def solve(self, remainCols: int, prevIdx: int, m: int) -> int:
        if remainCols == 0:
            return 1
        if self.dp[remainCols][prevIdx] != -1:
            return self.dp[remainCols][prevIdx]
        ways = 0
        prevState = self.columnStates[prevIdx]
        for i in range(len(self.columnStates)):
            if i == prevIdx:
                continue
            currState = self.columnStates[i]
            valid = all(prevState[j] != currState[j] for j in range(m))
            # valid = True
            # for j in range(m):
            #     if prevState[j] == currState[j]:
            #         valid = False
            if valid:
                ways = (ways + self.solve(remainCols - 1, i, m)) % self.MOD
        self.dp[remainCols][prevIdx] = ways
        return ways

    def colorTheGrid(self, m: int, n: int) -> int:
        self.generateColumnStates("", '#', 0, m)
        totalStates = len(self.columnStates)
        self.dp = [[-1] * (totalStates + 1) for _ in range(n + 1)]
        result = 0
        for i in range(totalStates):
            result = (result + self.solve(n - 1, i, m)) % self.MOD
        return result

File: Painting a Grid With Three Different Colors/Painting a Grid With Three Different Colors.cpp
// Solution for Painting a Grid With Three Different Colors in CPP

class Solution {
public:
    vector<string> columnStates;
    int MOD = 1e9 + 7;
    vector<vector<int>> dp;

    void generateColumnStates(string curr, char prevChar, int l, int m) {
        if (l == m) {
            columnStates.push_back(curr);
            return;
        }
        for (char currColor: {'R', 'G', 'B'}) {
            if (currColor == prevChar)
                continue;
            generateColumnStates(curr + currColor, currColor, l + 1, m);
        }
    }

    int solve(int remainCols, int prevIdx, int m) {
        if (remainCols == 0) {
            return 1;
        }
        if (dp[remainCols][prevIdx] != -1) {
            return dp[remainCols][prevIdx];
        }
        int ways = 0;
        string prevState = columnStates[prevIdx];
        for (int i = 0; i < columnStates.size(); i++) {
            if (i == prevIdx)
                continue;
            string currState = columnStates[i];
            bool valid = true;
            for (int j = 0; j < m; j++) {
                if (prevState[j] == currState[j]) {
                    valid = false;
                }
            }
            if (valid) {
                ways = (ways + solve(remainCols - 1, i, m)) % MOD;
            }
        }
        return dp[remainCols][prevIdx] = ways;
    }

    int colorTheGrid(int m, int n) {
        // generateColumnStates(current string, previous character, length, m (row size)) 
        generateColumnStates("", '#', 0, m);
        int totalStates = columnStates.size();
        dp = vector<vector<int>>(n + 1, vector<int>(totalStates + 1, -1));
        int result = 0;
        for (int i = 0; i < columnStates.size(); i++) {
            result = (result + solve(n - 1, i, m)) % MOD;
        }
        return result;
    }
};
